kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetricsSolverListener.java

```diff
@@ -42,7 +42,13 @@ public void solvingStarted(DefaultSolverScope solverScope) {
     @Override
     public void solvingEnded(DefaultSolverScope solverScope) {
         metrics.getOptaPlannerSolverCount().dec();
+        metrics.getOptaPlannerSolverScoreCalculationSpeed().labels(solverId).observe(getScoreCalculationSpeed(solverScope));
         timer.observeDuration();
     }
 
+    private double getScoreCalculationSpeed(DefaultSolverScope solverScope) {
+        long timeMillisSpent = solverScope.calculateTimeMillisSpentUpToNow();
+        // Avoid divide by zero exception on a fast CPU
+        return solverScope.getScoreCalculationCount() * 1000L / (timeMillisSpent == 0L ? 1L : timeMillisSpent);
```

Comment:
This formula will give results similar to `1M / 0 = 1M` (for 1 million calculations in "no time"), which is incorrect in my opinion. Such solver would **appear to be slower** than solver with same configuration running on a slower machine, where for example `1M / 0.1 = 10M`.

My proposal is to rewrite the formula so that it gives `<anything> / 0 = Double.MAX_VALUE`. Do you agree?

Reply:
@yurloc TBH, I have not given this enough thought - I just rewrote what I found in the original method implementation at https://github.com/kiegroup/optaplanner/blob/master/optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/scope/DefaultSolverScope.java#L210

Reply:
Interesting. Maybe @ge0ffrey had good reasons to write it that way. Without explanation it doesn't make sense to me, though.

Reply:
@yurloc may I ask how you got the `1M / 0.1`? The divisor is a `long` type. Not sure what's better - Double.MAX_VALUE is not precise (no computation takes 0 time), nor making the divisor `1L` is. If we don't want to measure nanoseconds.

Reply:
@rsynek `0.1` was meant to represent a very small amount of time larger than zero. I didn't realize we're using milliseconds. In that case, `1` has the same function but I'd still argue that `x / 0 = double.max` is much more precise approximation than `x / 0 = x`.

Reply:
@yurloc That upstream optaplanner-code looks fishy indeed on first sight. This code should 100% match that upstream code, after any change we agree. Some notes:
- Because timeMillisSpent is a long, so it's indeed either 0L or 1L, never 0.1, so that reasoning is void.
- A computer's internal clock has an accuracy, depending on the model and the OS. For example I used to see an accuracy of 15 milliseconds a lot 15 years ago on Windows. That means that something that takes 8 milliseconds will randomly resolve to 0ms or 15ms timeMillisSpent.
- If the timeMillisSpent resolves to 0ms, this doesn't mean that the score calculation was infinitely fast. It's not positive infinity.
- In the same reasoning, Double.MAX_VALUE is going to be far more off than just taking the score calculation count per second (so the score calc multiplied by 1000L and divided by 1L).
-- For example: we have a calc count of 36 over 3 milliseconds. That should end up as **12 per second**.
--- But the clock gave back 0 milliseconds due to the 15ms accuracy. What is the most correct number?
---- 36 * 1000L / 0L = Positive infinity per second
---- Double.MAX_VALUE per second, so somewhere around "2 to the power 1024"  per second
---- 36 * 1000L / 1L = 36 per second
--- The last one is the the closest.

So, given this reasoning, **I believe the upstream optaplanner-core is correct**. Wdyt?

Reply:
Upstream code changes I am lumping on the PR for the new SolverListener:

```

        // Avoid divide by zero exception on a run faster than the clock accuracy (which can be up to 15 milliseconds)
        // In that case, the 0 milliseconds gets rounded up to 1 millisecond,
        // which is closer to the correct result than positive infinity or MAX_VALUE.
        // Ideally, we'd replace that 0 by the half of the clock accuracy instead, but that information is unavailable.
        return getScoreCalculationCount() * 1000L / (timeMillisSpent == 0L ? 1L : timeMillisSpent);
```

The last line is unchanged.

Reply:
I'll add this comment to the current code on this branch

Reply:
@ge0ffrey Thanks for detailed explanation. It makes much more sense now. I don't think the formula is *correct*. It's a good approximation but it is essentially wrong because we *don't know* the actual value. On a second thought, `Double.MAX_VALUE` isn't a good idea because it's indeed a worse approximation.

There's another option that's better than `Double.MAX_VALUE` and it is `Double.NaN`. It's the most *correct* result. I would say it's definitely more correct than `x / 1`. The problem is that it lacks any information about `x`, which makes it a worse approximation than `x / 1`.

| Formula | Correctness score | Approximation score |
| --- | --- | --- |
| x / 1 | Poor | Good |
| NaN | Good | Nil |

So, I can now see both approaches as acceptable. The choice depends on context. I would go for `NaN` since I suppose the result isn't used for further computation, it's just displayed to the user, and I'd like to be made aware of the exceptional (and rare) condition. In practice, since there's no real reason for a change, let's stick with the old formula :slightly_smiling_face:.

Reply:
@vblagoje I created a function to concentrate this conversion, as Prometheus suggests as good practice to store time-related metrics as seconds, see https://prometheus.io/docs/practices/naming/#base-units and https://github.com/kiegroup/droolsjbpm-integration/pull/1799/files#diff-36b4ccc121dfe965799ba3292507b0eaR45 should we make sure we both share the same approach?